File: src/check-fields.ts
/**
 * Check fields for a specific issue to verify the Ops Team Designation field.
 * This is useful for debugging issues with field values.
 * 
 * Usage:
 *   ts-node src/check-fields.ts --issue-key TS-24130
 */

import * as dotenv from 'dotenv';
import axios from 'axios';
import * as yargs from 'yargs';

dotenv.config();

interface FieldMapping {
  [fieldName: string]: string;
}

interface CliArguments {
  'issue-key': string;
}

const fieldIds: FieldMapping = {
  'Ops Team Designation': 'customfield_10249',
  'Customer(s)': 'customfield_10485',
  'Type of Request': 'customfield_10617'
};

function extractTextFromADF(adfDoc: any): string {
  const text: string[] = [];
  
  const processContent = (content: any): void => {
    if (Array.isArray(content)) {
      content.forEach(item => processContent(item));
    } else if (content && typeof content === 'object') {
      if (content.type === 'text') {
        text.push(content.text || '');
      }
      if (content.content) {
        processContent(content.content);
      }
    }
  };
  
  processContent(adfDoc.content || []);
  return text.join('');
}

async function main(): Promise<void> {
  const argv = yargs
    .option('issue-key', {
      type: 'string',
      required: true,
      description: 'Issue key to check'
    })
    .help()
    .argv as CliArguments;

  // Initialize axios with authentication
  const axiosInstance = axios.create({
    baseURL: process.env.JIRA_URL || 'https://your-domain.atlassian.net',
    auth: {
      username: process.env.JIRA_USERNAME || '',
      password: process.env.JIRA_API_TOKEN || ''
    },
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'application/json'
    }
  });

  try {
    // Get issue details
    const response = await axiosInstance.get(`/rest/api/3/issue/${argv['issue-key']}`);
    const issue = response.data;

    if (!issue) {
      console.log(`Issue ${argv['issue-key']} not found`);
      return;
    }

    const fields = issue.fields || {};

    // Print basic issue info
    console.log(`\n=== Issue Details for ${argv['issue-key']} ===`);
    console.log(`Summary: ${fields.summary || 'N/A'}`);
    console.log(`Status: ${fields.status?.name || 'N/A'}`);
    console.log(`Issue Type: ${fields.issuetype?.name || 'N/A'}`);
    console.log(`Reporter: ${fields.reporter?.displayName || 'N/A'}`);
    console.log(`Reporter Email: ${fields.reporter?.emailAddress || 'N/A'}`);

    // Print specific fields we care about
    console.log('\n=== Required Fields ===');
    for (const [fieldName, fieldId] of Object.entries(fieldIds)) {
      const fieldValue = fields[fieldId];
      console.log(`${fieldName}: `, { endLine: false });

      if (fieldValue === null || fieldValue === undefined) {
        console.log('NOT SET');
      } else if (Array.isArray(fieldValue)) {
        const values = fieldValue.map(item => {
          if (typeof item === 'object' && item !== null && 'value' in item) {
            return item.value;
          }
          return String(item);
        });
        console.log(values.join(', '));
      } else if (typeof fieldValue === 'object' && fieldValue !== null && 'value' in fieldValue) {
        console.log(fieldValue.value);
      } else {
        console.log(fieldValue);
      }
    }

    // Print description (first 200 chars) for keyword analysis
    const description = fields.description;
    if (description) {
      let descText = '';
      
      if (typeof description === 'string') {
        // Handle plain text description
        descText = description;
      } else if (typeof description === 'object' && description.content) {
        // Handle Atlassian Document Format
        descText = extractTextFromADF(description);
      }
      
      if (descText) {
        console.log(`\nDescription (first 200 chars): ${descText.substring(0, 200)}...`);
      }
    }

    // Print raw field values for debugging
    console.log('\n=== Raw Field Values ===');
    for (const [fieldName, fieldId] of Object.entries(fieldIds)) {
      const fieldValue = fields[fieldId];
      console.log(`${fieldName} (${fieldId}): ${JSON.stringify(fieldValue, null, 2)}`);
    }

  } catch (error: any) {
    console.error('Error:', error.message);
    if (error.response) {
      console.error('Response status:', error.response.status);
      console.error('Response data:', error.response.data);
    }
  }
}

function console_log_no_newline(text: string): void {
  process.stdout.write(text);
}

// Extend console with a method to print without newline
declare global {
  interface Console {
    log(message?: any, options?: { endLine: boolean }): void;
  }
}

if (require.main === module) {
  main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

File: check_fields.py
"""
Check fields for a specific issue to verify the Ops Team Designation field.
This is useful for debugging issues with field values.

Usage:
    python check_fields.py --issue-key TS-24130
"""

import os
import argparse
import json
from atlassian import Jira
from dotenv import load_dotenv

def main():
    parser = argparse.ArgumentParser(description='Check specific fields of an issue')
    parser.add_argument('--issue-key', type=str, required=True, help='Issue key to check')
    args = parser.parse_args()
    
    # Load environment variables
    load_dotenv()
    
    # Initialize Jira client
    jira = Jira(
        url=os.getenv('JIRA_URL', 'https://certifyos.atlassian.net'),
        username=os.getenv('JIRA_USERNAME', ''),
        password=os.getenv('JIRA_API_TOKEN', '')
    )
    
    # Get issue details
    issue = jira.issue(args.issue_key)
    if not issue:
        print(f"Issue {args.issue_key} not found")
        return
    
    # Get fields
    fields = issue.get('fields', {})
    
    # Field IDs to check
    field_ids = {
        'Ops Team Designation': 'customfield_10249',
        'Customer(s)': 'customfield_10485',
        'Type of Request': 'customfield_10617'
    }
    
    # Print basic issue info
    print(f"\n=== Issue Details for {args.issue_key} ===")
    print(f"Summary: {fields.get('summary', 'N/A')}")
    print(f"Status: {fields.get('status', {}).get('name', 'N/A')}")
    print(f"Issue Type: {fields.get('issuetype', {}).get('name', 'N/A')}")
    print(f"Reporter: {fields.get('reporter', {}).get('displayName', 'N/A')}")
    print(f"Reporter Email: {fields.get('reporter', {}).get('emailAddress', 'N/A')}")
    
    # Print specific fields we care about
    print("\n=== Required Fields ===")
    for field_name, field_id in field_ids.items():
        field_value = fields.get(field_id)
        print(f"{field_name}: ", end="")
        
        if field_value is None:
            print("NOT SET")
        elif isinstance(field_value, list):
            values = []
            for item in field_value:
                if isinstance(item, dict) and 'value' in item:
                    values.append(item['value'])
                else:
                    values.append(str(item))
            print(', '.join(values))
        elif isinstance(field_value, dict) and 'value' in field_value:
            print(field_value['value'])
        else:
            print(field_value)
    
    # Print description (first 200 chars) for keyword analysis
    description = fields.get('description', 'N/A')
    if isinstance(description, dict) and 'content' in description:
        # Handle Atlassian Document Format
        desc_text = extract_text_from_adf(description)
        if desc_text:
            print(f"\nDescription (first 200 chars): {desc_text[:200]}...")
    elif isinstance(description, str):
        # Handle plain text description
        print(f"\nDescription (first 200 chars): {description[:200]}...")
    
    # Print raw field values for debugging
    print("\n=== Raw Field Values ===")
    for field_name, field_id in field_ids.items():
        field_value = fields.get(field_id)
        print(f"{field_name} ({field_id}): {json.dumps(field_value, indent=2)}")

def extract_text_from_adf(adf_doc):
    """Extract plain text from Atlassian Document Format"""
    text = []
    
    def process_content(content):
        if isinstance(content, list):
            for item in content:
                process_content(item)
        elif isinstance(content, dict):
            if content.get('type') == 'text':
                text.append(content.get('text', ''))
            if 'content' in content:
                process_content(content['content'])
    
    process_content(adf_doc.get('content', []))
    return ''.join(text)

if __name__ == "__main__":
    main()
